### DISCUSSION TITLE
Controlling frame rate of an Animation?

### DISCUSSION THREAD

**Discussion by thertzelle** [visgl/deck.gl#5231] [Q&A]
It it possible to control the frame rate of an animation transition? We have a few use cases where drawing at 60 fps for an animation is not needed and is too performance heavy for our application. I've done a little searching, but cannot find a direct way to do this, so I was looking for some suggestions?

Thank you.

**Comment by Pessimistress** [comment]
What kind of animation are you talking about? Attribute transition? Viewport transition? Data updates?

**Comment by thertzelle** [comment]
Thank you for your reply and questions. I'm hoping to animate one to six ScatterplotLayer circles to represent a real-time wave traveling from a point. Last night I was using an update trigger to change the radius (with a transition on the radius). Then I saw the notes in (https://deck.gl/docs/developer-guide/performance) but it wasn't exactly about animation, so I switched it to update the radius and the performance improved slightly, but my final temp solution is to extremely manually update the radius without animation (looks a bit stuttery compared to the layer transition animation. Any tips here? The radius changes perhaps every second due to the wave being re-calculated based on speed.

**Comment by Pessimistress** [comment]
If you don't have millions of circles, it's unlikely the rendering that is causing the performance issue. I can't tell what's wrong from your description. Please create a Codepen/CodeSandbox if you need further advice.

**Comment by chrisgervang** [comment]
I haven't done this before so apologies in advance if I assert something incorrectly, but I have some thoughts that might be helpful in achieving a "low frame rate" animation. Deck's framerate is ultimately fluid and tied to `requestAnimationFrame` - in other words, I think of it as a coincidence that deck runs at 60 fps (this is fundamentally controlled by the browser and there is a discussion about [working around this](https://dev.to/uwutrinket/chrome-unlimited-frame-rate-gl2)). This means you can utilize a ~60 fps tick to animate with. 

If guaranteeing the duration of your animation is important, then comparing the "wall clock" (`Performance.now()`) on every tick and only drawing if a desired "interval < 60Hz" has passed is the way to go. There's a related [stack overflow](https://stackoverflow.com/a/19772220) describing this concept, but **I think there's a better built-in way** to do all of this via `AnimationLoop`.

For supporting animations, Deck attaches a luma.gl [`AnimationLoop`](https://github.com/visgl/luma.gl/blob/master/modules/engine/src/lib/animation-loop.js) and plays it when you do `new Deck({ _animate: true })`. The built-in animation engine is great for attaching keyframed effects (and supports looping too via "channels"). I'd recommend looking into the [Timeline](https://luma.gl/docs/api-reference/engine/animation/timeline) and [KeyFrame](https://luma.gl/docs/api-reference/engine/animation/key-frames) docs for ideas. The [animation-timeline-rfc](https://github.com/visgl/luma.gl/blob/master/dev-docs/RFCs/v7.1/animation-timeline-rfc.md) and [keyframe-rfc](https://github.com/visgl/luma.gl/blob/master/dev-docs/RFCs/v7.1/keyframe-rfc.md) are great for developing a mental model for "channels".

If you describe some more of your use case I may have some more ideas on how to apply this.

**Comment by ibgreen** [comment]
One option is to use a `_customRender` function in deck.gl https://github.com/visgl/deck.gl/blob/master/modules/core/src/lib/deck.js#L348

You could check the time compared to last render and return if too soon, letting you cap at say 30, or 15 frames per second (@chrisgervang 's second paragraph above describes how to do this). 

However, I am not sure if you can determine if rendering is happening because of a transition, so I suppose this would throttle the frame rate of all deck.gl rendering. including user interaction, panning, zooming etc. But if you are having perf issues in transition, you probably have them everywhere.

**Comment by zakjan** [comment]
Using `_customRender` for limiting _animate FPS works great. For anyone interested, here is my snippet:

```
let lastRenderTime = 0;
const deckgl = new Deck({
  ...
  _customRender: redrawReason => {
    const fps = 30;
    const fpsInterval = 1000 / fps;
    const now = Date.now();
    const elapsed = now - lastRenderTime;
    if (elapsed > fpsInterval) {
      lastRenderTime = now;
      deckgl._drawLayers(redrawReason);
    }
  },
});
```

I'm wondering, is there last render time stored somewhere in deckgl state, so that I don't need to store it myself on side?
